src/optimizer/filter_pushdown.cpp

```diff
@@ -1,17 +1,89 @@
 #include "duckdb/optimizer/filter_pushdown.hpp"
 
 #include "duckdb/optimizer/filter_combiner.hpp"
+#include "duckdb/optimizer/optimizer.hpp"
+#include "duckdb/planner/expression_iterator.hpp"
 #include "duckdb/planner/operator/logical_filter.hpp"
 #include "duckdb/planner/operator/logical_join.hpp"
 #include "duckdb/planner/operator/logical_window.hpp"
-#include "duckdb/optimizer/optimizer.hpp"
+#include "duckdb/planner/operator/logical_projection.hpp"
+#include "duckdb/planner/operator/logical_comparison_join.hpp"
 
 namespace duckdb {
 
 using Filter = FilterPushdown::Filter;
 
-FilterPushdown::FilterPushdown(Optimizer &optimizer, bool convert_mark_joins)
+static unordered_set<idx_t> GetMarkJoinIndexes(LogicalOperator &plan, unordered_set<idx_t> &table_bindings) {
+	unordered_set<idx_t> projected_mark_join_indexes;
+	switch (plan.type) {
+	case LogicalOperatorType::LOGICAL_COMPARISON_JOIN: {
+		auto &join = plan.Cast<LogicalComparisonJoin>();
+		if (join.join_type != JoinType::MARK) {
+			break;
+		}
+		// if the projected table bindings include the mark join index,
+		if (table_bindings.find(join.mark_index) != table_bindings.end()) {
+			projected_mark_join_indexes.insert(join.mark_index);
+		}
+		break;
+	}
+	// you need to store table.column index.
+	// if you get to a projection, you need to change the table_bindings passed so they reflect the
+	// table index of the original expression they originated from.
+	case LogicalOperatorType::LOGICAL_PROJECTION: {
+		// when we encounter a projection, replace the table_bindings with
+		// the tables in the projection
+		auto plan_bindings = plan.GetColumnBindings();
+		auto &proj = plan.Cast<LogicalProjection>();
+		auto proj_bindings = proj.GetColumnBindings();
+		unordered_set<idx_t> new_table_bindings;
+		for (auto &binding : proj_bindings) {
+			auto col_index = binding.column_index;
+			auto &expr = proj.expressions.at(col_index);
+			vector<ColumnBinding> bindings_to_keep;
+			ExpressionIterator::EnumerateExpression(expr, [&](Expression &child) {
```

Comment:
The `ExpressionIterator` only iterates one level - I'm not sure if that is sufficient here. 

The following query still seems to throw an internal exception:

```sql
CREATE OR REPLACE TABLE BaseData AS (
        SELECT
          '10' AS my_key,
          '20' AS parent_key,
          '30' AS payload,
          '40' as foo,
          '50' as foo2,
          '60' as foo3
    );
D WITH
      Example AS (
          SELECT
              c.my_key,
              (c.parent_key IN (SELECT my_key FROM BaseData)) AND c.my_key>'10' AS parentExists,
              p.my_key IS NOT NULL AS parentExists2,
          FROM BaseData AS c
          LEFT JOIN BaseData AS p ON c.parent_key = p.my_key
      )
      SELECT *
      FROM Example
      WHERE parentExists;
```

The `LogicalOperatorVisitor` can do recursive visiting of expressions that should likely be used here.